#!/bin/bash
#PBS -q condo
#PBS -N download_igm
#PBS -l nodes=1:ppn=1
#PBS -l walltime=6:00:00
#PBS -V
#PBS -m abe
#PBS -A epigen-group

# -v pass parameters: ftp_addr,user,pass, flowcell_id


############################################################
# args: $flowcell_id,  $ftp_addr $user $password 
############################################################
cmd="source activate django;python \$(which updateRunStatus.py) -s '1' -f $flowcell_id"
ssh zhc268@epigenomics.sdsc.edu $cmd

#ftp_addr=ftp://igm-storage1.ucsd.edu/190215_K00180_0755_BH2WKMBBXY_SR75_Combo/
#ftp_addr=$1
#user=$2
#pass=$3
#flowcell_id=$4
############################################################
# transfer data 
############################################################
cd  /home/zhc268/data/seqdata
cmd="wget  -r -l 10 -q --user $user --password $pass  ${ftp_addr}/*"
echo $cmd 
echo $cmd | bash

############################################################
# get reads cnt 
############################################################
fastq_dir="/projects/ps-epigen/seqdata/${ftp_addr##*//}"
reads_cnt_file="${fastq_dir}/reads_cnt.tsv"
>$reads_cnt_file

cd $fastq_dir

## not demultiplexed 
##if [ ! -f *_R1_*.fastq.gz ] 
##then
##    qsub -v flowcell_id=,run_dir=${fastq_dir}
##fi
## TODO 

## demultiplexed 
libs=($(ls -1  *_R1_*.fastq.gz | sed 's/_S[0-9]*.*//g'|grep -v Undet|sort|uniq))
if ls  *_R3_*fastq.gz 1> /dev/null 2>&1; then IS_10x=T;fi
if ls  *_R2_*fastq.gz 1> /dev/null 2>&1; then IS_PE=T;fi

## count reads & make sybolic links for each lib
for lib in ${libs[@]}
do
    echo $lib
    nlane=$(ls -1 ${lib}_S*_R1*.fastq.gz |wc -l)
    ## handle PE  
    if [[ ! -z  $IS_PE && $nlane -gt 1 ]] ## more than 1 lane
    then
        echo -e "cat  ${fastq_dir}/${lib}*R1*.fastq.gz > /home/zhc268/data/seqdata/${lib}_R1.fastq.gz"|bash
        if [ ! -z $IS_10x ] ## 10x
        then
             mkdir -p /home/zhc268/data/seqdata/${lib}
             ln -sf ${fastq_dir}/${lib}*.fastq.gz /home/zhc268/data/seqdata/${lib}/
             echo -e "cat  ${fastq_dir}/${lib}*R3*.fastq.gz > /home/zhc268/data/seqdata/${lib}_R2.fastq.gz"|bash
        elif [ -z $IS_10x ] ## bulk
        then
             echo -e "cat  ${fastq_dir}/${lib}*R2*.fastq.gz > /home/zhc268/data/seqdata/${lib}_R2.fastq.gz"|bash
        fi
        nreads=$(zcat /home/zhc268/data/seqdata/${lib}_R1.fastq.gz | wc -l ) && echo -e "$lib\t$[nreads/4]" >> $reads_cnt_file        
    elif [[ ! -z  $IS_PE && $nlane -eq 1 ]] # only one lane 
    then
        ln -sf ${fastq_dir}/${lib}*R1*.fastq.gz /home/zhc268/data/seqdata/${lib}_R1.fastq.gz
        if [ ! -z $IS_10x ] ## 10x
        then
             mkdir -p /home/zhc268/data/seqdata/${lib}
             ln -sf ${fastq_dir}/${lib}*.fastq.gz /home/zhc268/data/seqdata/${lib}/
             ln -sf ${fastq_dir}/${lib}*R3*.fastq.gz  /home/zhc268/data/seqdata/${lib}_R2.fastq.gz 
        elif [ -z $IS_10x ] ## bulk 
        then
             ln -sf  ${fastq_dir}/${lib}*R2*.fastq.gz  /home/zhc268/data/seqdata/${lib}_R2.fastq.gz 
        fi
        nreads=$(zcat /home/zhc268/data/seqdata/${lib}_R1.fastq.gz | wc -l ) && echo -e "$lib\t$[nreads/4]" >> $reads_cnt_file        
    fi

    ## handle  SE 
    if [[ -z $IS_PE && $nlane -gt 1 ]] #handle SE & more than 1 lane
    then
        cat  ${fastq_dir}/${lib}*R1*.fastq.gz > /home/zhc268/data/seqdata/${lib}.fastq.gz
        nreads=$(zcat /home/zhc268/data/seqdata/${lib}.fastq.gz | wc -l ) && echo -e "$lib\t$[nreads/4]" >> $reads_cnt_file        
    elif [[ -z $IS_PE && $nlane -eq 1 ]] #handle SE & more than 1 lane
    then
         ln -sf ${fastq_dir}/${lib}*R1*.fastq.gz /home/zhc268/data/seqdata/${lib}.fastq.gz
         nreads=$(zcat /home/zhc268/data/seqdata/${lib}.fastq.gz | wc -l ) && echo -e "$lib\t$[nreads/4]" >> $reads_cnt_file        
    fi
done
wait

############################################################
# update status to finish or warning @ VM
############################################################
cmd="source activate django; python \$(which updateReadsNumberPerRun.py) -f $flowcell_id -i $reads_cnt_file; python \$(which updateRunReads.py) -f $flowcell_id"
ssh zhc268@epigenomics.sdsc.edu $cmd

