#!/bin/bash
#PBS -q condo
#PBS -N django_10xMkfastqs
#PBS -l nodes=1:ppn=16
#PBS -l walltime=8:00:00
#PBS -o /home/zhc268/logs/run_10xMkfastqs.out
#PBS -e /home/zhc268/logs/run_10xMkfastqs.err
#PBS -V
#PBS -m abe
#PBS -A epigen-group

# -v pass parameters: flowcell_id, run_dir
############################################################
## update status to job submitted @VM
############################################################
cmd="source activate django;python \$(which updateRunStatus.py) -s '1' -f $flowcell_id"
ssh zhc268@epigenomics.sdsc.edu $cmd



############################################################
## run bcl2fastq @TSCC
############################################################
## delete last demux info 
[[ -d ${run_dir}/${flowcell_id} ]] && rm -r ${run_dir}/${flowcell_id}  # delete existing folder
[[ -f ${run_dir}/__${flowcell_id}.mro ]] && rm ${run_dir}/__${flowcell_id}.mro  # delete existing folder

## prepare variables & directories
out_dir=${run_dir}/Data/Fastqs/
[[ -d $out_dir ]] && rm -r $out_dir # delete existing folder
mkdir -p $out_dir


## prepair reads_cnt_file
reads_cnt_file="${out_dir}/reads_cnt.tsv"
>$reads_cnt_file

## samplesheet
samplesheet="${out_dir}/SampleSheet.csv"

cd $run_dir
export PATH=/projects/ps-epigen/software/cellranger-atac-1.0.1/:$PATH
cellranger-atac mkfastq --run=$run_dir --output-dir=$out_dir --localcores=16 --csv $samplesheet --qc 


# create simlink to the seqdata folder & cat reads 
for i in $(awk -v FS=',' '(NR>1){print $2}' $samplesheet)
do
    find $(pwd) -name "$i" -type d -exec ln -s {} ~/data/seqdata/ \; ; 
    cat ~/data/seqdata/${i}/${i}*R3_*fastq.gz > ~/data/seqdata/${i}_R2.fastq.gz & sleep 1
    cat ~/data/seqdata/${i}/${i}*R1_*fastq.gz > ~/data/seqdata/${i}_R1.fastq.gz & sleep 1
done

wait 
